syntax = "proto3";

package raft;

service RaftService {
  rpc RequestVote (RequestVoteArgs) returns (RequestVoteReply) {}
  rpc AppendEntries (AppendEntriesArgs) returns (AppendEntriesReply) {}
  rpc SubmitCommand (CommandRequest) returns (CommandReply) {}
}

message LogEntry {
  uint64 term = 1;
  string command = 2;
}

message RequestVoteArgs {
  uint64 term = 1;
  int32 candidate_id = 2;    // Changed to int32
  uint64 last_log_index = 3;
  uint64 last_log_term = 4;
}

message RequestVoteReply {
  uint64 term = 1;
  bool vote_granted = 2;
}

message AppendEntriesArgs {
  uint64 term = 1;
  int32 leader_id = 2;       // Changed to int32
  uint64 prev_log_index = 3;
  uint64 prev_log_term = 4;
  repeated LogEntry entries = 5;
  uint64 leader_commit = 6;
}

message AppendEntriesReply {
  uint64 term = 1;
  bool success = 2;
}

message CommandRequest {
  string command = 1;
}

message CommandReply {
  bool success = 1;
  int32 leader_id = 2;
  string message = 3;
}